File: src/Java/BehavioralPatterns/Mediator/CommunicationHub/src/Participants/Sensor.java
package Participants;

import java.util.Random;

public class Sensor extends Participant {
    private final String id;
    private Object lastValue;

    public Sensor(String id) {
        this.id = id;
    }

    public String getId() {
        return id;
    }

    public Object getLastValue() {
        return lastValue;
    }

    public void setLastValue(Object lastValue) {
        this.lastValue = lastValue;
    }

    @Override
    public void receive(Participant sender, Object args) {
        if (args.equals("measure")) {
            lastValue = new Random().nextDouble();
            getMediator().notify(this, lastValue);
        }
    }

    public void valueChanged(Object value) {
        lastValue = value;
        getMediator().notify(this, lastValue);
    }

    @Override
    public String toString() {
        return "Sensor(" + id + ")";
    }
}

File: src/JavaScript/BehavioralPatterns/Mediator/CommunicationHub/participants/sensor.js

const Participant = require("./participant");

class Sensor extends Participant {
  constructor(id) {
    super();
    this.id = id;
    this.lastValue = null;
  }

  get id() {
    return this._id;
  }

  set id(value) {
    this._id = value;
  }

  get lastValue() {
    return this._lastValue;
  }

  set lastValue(value) {
    this._lastValue = value;
  }

  receive(sender, args) {
    if (args === "measure") {
      this.lastValue = Math.random();
      this.mediator.notify(this, this.lastValue);
    }
  }

  valueChanged(value) {
    this.lastValue = value;
    this.mediator.notify(this, this.lastValue);
  }

  toString() {
    return `Sensor(${this.id})`;
  }
}

module.exports = Sensor;